File: src/connectors/googleTrends.connector.js
import googleTrends from 'google-trends-api';
import logger from '../utils/logger.js';
import { getDateRange } from '../utils/dates.js';

/**
 * Detectar si estamos en modo test
 */
const IS_TEST_MODE = process.env.NODE_ENV === 'test';

/**
 * Mock data generator para tests (evita llamadas a Google Trends API)
 */
function generateMockTimeSeries(keyword, baselineDays) {
  const totalDays = baselineDays + 1;
  const series = [];
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - baselineDays);

  const seed = keyword.split('').reduce((acc, char) => acc + char.charCodeAt(0), 0);
  
  for (let i = 0; i < totalDays; i++) {
    const date = new Date(startDate);
    date.setDate(date.getDate() + i);
    
    const dayOffset = i / totalDays;
    const baseValue = 30 + (seed % 40);
    const trend = Math.sin(dayOffset * Math.PI * 4) * 20;
    const noise = ((seed * (i + 1)) % 30) - 15;
    
    const value = Math.max(0, Math.min(100, Math.round(baseValue + trend + noise)));
    
    series.push({
      date: date.toISOString().split('T')[0],
      value
    });
  }

  return series;
}

function generateMockByCountry(keyword, country) {
  const seed = keyword.split('').reduce((acc, char) => acc + char.charCodeAt(0), 0);
  const countries = ['MX', 'CR', 'ES'];
  
  return countries.map(countryCode => {
    let value;
    if (countryCode === country) {
      value = 80 + (seed % 21);
    } else {
      const offset = countryCode.charCodeAt(0) + countryCode.charCodeAt(1);
      value = Math.max(0, Math.min(79, (seed + offset) % 80));
    }
    
    return { country: countryCode, value };
  }).sort((a, b) => b.value - a.value);
}

/**
 * Google Trends Connector - REAL API IMPLEMENTATION
 * 
 * Uses google-trends-api npm package to fetch real data from Google Trends.
 * 
 * IMPORTANT: Google may block requests if:
 * - Too many requests in short time (rate limiting)
 * - Unusual traffic patterns detected
 * 
 * Recommendations:
 * - Use delays between requests (2-5 seconds)
 * - Implement exponential backoff on errors
 * - Consider caching results for 12-24 hours
 * - For high-volume production, use SerpAPI instead
 * 
 * TEST MODE: When NODE_ENV=test, returns mock data instead of calling Google API
 */
class GoogleTrendsConnector {
  constructor() {
    this.maxRetries = parseInt(process.env.GOOGLE_TRENDS_MAX_RETRIES || '3', 10);
    this.retryDelay = parseInt(process.env.GOOGLE_TRENDS_RETRY_DELAY_MS || '5000', 10);
    this.requestDelay = parseInt(process.env.GOOGLE_TRENDS_REQUEST_DELAY_MS || '4000', 10);
    this.concurrency = parseInt(process.env.GOOGLE_TRENDS_CONCURRENCY || '1', 10);
    
    // Lock para garantizar concurrencia 1
    this.requestLock = null;
    this.lockQueue = [];
    
    if (IS_TEST_MODE) {
      logger.info('🧪 Google Trends Connector - MOCK MODE (No API calls will be made)');
    } else {
      logger.info({ 
        maxRetries: this.maxRetries,
        retryDelay: this.retryDelay,
        requestDelay: this.requestDelay,
        concurrency: this.concurrency
      }, 'Google Trends Connector initialized with REAL API - MVP Configuration');
    }
  }

  /**
   * Fetch complete trend data: time series + regional comparison
   */
  async fetchComplete(keyword, country, windowDays, baselineDays) {
    // Si estamos en modo test, retornar mock data inmediatamente
    if (IS_TEST_MODE) {
      await new Promise(resolve => setTimeout(resolve, 5)); // Simular delay mínimo
      return {
        timeSeries: generateMockTimeSeries(keyword, baselineDays),
        byCountry: generateMockByCountry(keyword, country)
      };
    }

    // Adquirir lock antes de hacer request
    await this._acquireLock();

    try {
      logger.info({ keyword, country, windowDays, baselineDays }, 
        'Fetching complete trend data from Google Trends API');

      const { startDate, endDate } = getDateRange(windowDays, baselineDays);

      // Fetch time series with retry logic
      const timeSeries = await this._fetchWithRetry(() =>
        this._fetchTimeSeries(keyword, country, startDate, endDate)
      );

      // Add delay before second request to avoid rate limiting
      await this._delay(this.requestDelay);

      // Fetch country comparison (global view - last 12 months)
      const byCountry = await this._fetchWithRetry(() =>
        this._fetchByCountry(keyword)
      );

      logger.info({ keyword, country, dataPoints: timeSeries.length }, 
        'Successfully fetched Google Trends data');

      return {
        timeSeries,
        byCountry,
        source: 'google_trends',
        fetchedAt: new Date().toISOString()
      };
    } catch (error) {
      logger.error({ error: error.message, keyword, country }, 
        'Failed to fetch Google Trends data');
      throw error;
    } finally {
      // Siempre liberar el lock
      this._releaseLock();
    }
  }

  /**
   * Fetch time series data for a keyword in a specific country
   */
  async _fetchTimeSeries(keyword, country, startDate, endDate) {
    logger.info({ keyword, country, startDate, endDate }, 
      'Fetching time series from Google Trends');

    const results = await googleTrends.interestOverTime({
      keyword,
      startTime: new Date(startDate),
      endTime: new Date(endDate),
      geo: country
    });

    const parsed = JSON.parse(results);

    if (!parsed.default || !parsed.default.timelineData) {
      throw new Error('Invalid response format from Google Trends');
    }

    // Transform Google Trends format to our format
    const timeSeries = parsed.default.timelineData.map(point => ({
      date: this._formatTimestamp(point.time),
      value: point.value[0] || 0
    }));

    logger.info({ keyword, country, dataPoints: timeSeries.length }, 
      'Time series fetched successfully');

    return timeSeries;
  }

  /**
   * Fetch regional comparison for a keyword across MX, CR, ES
   * Note: Uses last 12 months (Google Trends default) for global comparison
   * This is intentionally different from time series date range
   */
  async _fetchByCountry(keyword) {
    logger.info({ keyword }, 'Fetching country comparison from Google Trends');

    try {
      // Single request for all countries (more efficient, less likely to be blocked)
      const data = await googleTrends.interestByRegion({
        keyword,
        startTime: new Date(Date.now() - 365 * 24 * 60 * 60 * 1000), // Last year
        resolution: 'COUNTRY'
        // No geo parameter = global comparison
      });

      const parsed = JSON.parse(data);
      const results = [];

      // Extract values for our target countries
      const targetCountries = ['MX', 'CR', 'ES'];
      
      if (parsed.default && parsed.default.geoMapData) {
        for (const countryCode of targetCountries) {
          const countryData = parsed.default.geoMapData.find(d => d.geoCode === countryCode);
          const value = countryData?.value?.[0] || 0;
          
          results.push({
            country: countryCode,
            value
          });

          logger.info({ country: countryCode, value }, 
            'Country data extracted successfully');
        }
      } else {
        // Fallback: return zeros if no data
        logger.warn('No geoMapData found, using zeros');
        targetCountries.forEach(code => {
          results.push({ country: code, value: 0 });
        });
      }

      // Sort by value descending
      results.sort((a, b) => b.value - a.value);

      logger.info({ keyword, countries: results.length }, 
        'Country comparison fetched successfully');

      return results;
    } catch (error) {
      logger.error({ error: error.message, keyword }, 
        'Failed to fetch country comparison');
      
      // Return zeros on error
      return ['MX', 'CR', 'ES'].map(code => ({ country: code, value: 0 }));
    }
  }

  /**
   * Wrapper for fetchByCountry to maintain API compatibility
   */
  async fetchByCountry(keyword) {
    return this._fetchByCountry(keyword);
  }

  /**
   * Retry wrapper with exponential backoff
   */
  async _fetchWithRetry(fn) {
    let lastError;

    for (let attempt = 1; attempt <= this.maxRetries; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error;
        
        // Detectar si Google devolvió HTML (bloqueo anti-bot)
        const isBlocked = error.message && (
          error.message.includes('Unexpected token') ||
          error.message.includes('is not valid JSON') ||
          error.message.includes('html') ||
          error.message.includes('DOCTYPE')
        );

        logger.warn({ 
          attempt, 
          maxRetries: this.maxRetries,
          error: error.message,
          isBlocked,
          nextDelay: attempt < this.maxRetries ? this.retryDelay * Math.pow(2, attempt - 1) : 0
        }, isBlocked ? 'Google blocked request (returned HTML), retrying...' : 'Request failed, retrying...');

        if (attempt < this.maxRetries) {
          // Exponential backoff: 5s, 10s, 20s
          const delay = this.retryDelay * Math.pow(2, attempt - 1);
          
          // Si fue bloqueado, agregar delay extra
          const extraDelay = isBlocked ? 3000 : 0;
          
          await this._delay(delay + extraDelay);
        }
      }
    }

    throw new Error(`Failed after ${this.maxRetries} attempts: ${lastError.message}`);
  }

  /**
   * Delay helper
   */
  _delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Adquirir lock para garantizar concurrencia 1
   */
  async _acquireLock() {
    return new Promise((resolve) => {
      if (!this.requestLock) {
        // No hay lock activo, adquirir inmediatamente
        this.requestLock = true;
        logger.debug('Lock acquired');
        resolve();
      } else {
        // Ya hay un request en curso, agregar a la cola
        logger.debug({ queueLength: this.lockQueue.length }, 'Waiting for lock...');
        this.lockQueue.push(resolve);
      }
    });
  }

  /**
   * Liberar lock y procesar siguiente en cola
   */
  _releaseLock() {
    if (this.lockQueue.length > 0) {
      // Hay requests esperando, dar lock al siguiente
      const next = this.lockQueue.shift();
      logger.debug({ remainingInQueue: this.lockQueue.length }, 'Lock passed to next in queue');
      next();
    } else {
      // No hay nadie esperando, liberar completamente
      this.requestLock = null;
      logger.debug('Lock released');
    }
  }

  /**
   * Format Google Trends timestamp to YYYY-MM-DD
   */
  _formatTimestamp(timestamp) {
    // Google Trends returns timestamps in seconds (sometimes as string)
    const ts = typeof timestamp === 'string' ? parseInt(timestamp) : timestamp;
    const date = new Date(ts * 1000);
    return date.toISOString().split('T')[0];
  }
}

export default new GoogleTrendsConnector();

File: app/connectors/google_trends_connector.py
"""Google Trends Connector using pytrends library."""
import os
import re
import time
import math
import random
import threading
from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple
import requests as _requests_lib  # kept for potential future proxy checks
from pytrends.request import TrendReq
from loguru import logger

# Global lock — serializes ALL pytrends requests across threads/greenlets
_GLOBAL_REQUEST_LOCK = threading.Lock()

# Detect test mode
IS_TEST_MODE = os.getenv('NODE_ENV') == 'test'

# User agents más realistas para evitar bloqueos
USER_AGENTS = [
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:123.0) Gecko/20100101 Firefox/123.0',
    'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.2.1 Safari/605.1.15',
]


def generate_mock_time_series(keyword: str, baseline_days: int) -> List[Dict]:
    """
    Generate mock time series data for testing.
    Deterministic based on keyword to ensure consistency.
    """
    total_days = baseline_days + 1
    series = []
    start_date = datetime.now() - timedelta(days=baseline_days)
    
    # Generate seed from keyword
    seed = sum(ord(char) for char in keyword)
    
    for i in range(total_days):
        date = start_date + timedelta(days=i)
        
        # Generate deterministic but varied values
        day_offset = i / total_days
        base_value = 30 + (seed % 40)
        trend = math.sin(day_offset * math.pi * 4) * 20
        noise = ((seed * (i + 1)) % 30) - 15
        
        value = max(0, min(100, round(base_value + trend + noise)))
        
        series.append({
            'date': date.strftime('%Y-%m-%d'),
            'value': value
        })
    
    return series


def generate_mock_by_country(keyword: str, country: str) -> List[Dict]:
    """
    Generate mock country comparison data for testing.
    Deterministic based on keyword to ensure consistency.
    """
    seed = sum(ord(char) for char in keyword)
    countries = ['MX', 'CR', 'ES']
    
    result = []
    for country_code in countries:
        if country_code == country:
            value = 80 + (seed % 21)
        else:
            offset = ord(country_code[0]) + ord(country_code[1])
            value = max(0, min(79, (seed + offset) % 80))
        
        result.append({'country': country_code, 'value': value})
    
    # Sort by value descending
    result.sort(key=lambda x: x['value'], reverse=True)
    return result


class GoogleTrendsConnector:
    """
    Google Trends Connector using pytrends library.
    
    Features:
    - Mock mode for testing (when NODE_ENV=test)
    - Retry logic with exponential backoff
    - Anti-blocking delays between requests
    - Concurrency control (locks)
    
    PyTrends is more stable than google-trends-api (Node.js),
    but Google can still block requests if too frequent.
    """
    
    def __init__(self):
        """Initialize connector with configuration from environment."""
        self.max_retries = int(os.getenv('GOOGLE_TRENDS_MAX_RETRIES', 5))
        self.retry_delay = int(os.getenv('GOOGLE_TRENDS_RETRY_DELAY_MS', 15000)) / 1000  # Convert to seconds
        self.request_delay = int(os.getenv('GOOGLE_TRENDS_REQUEST_DELAY_MS', 10000)) / 1000
        self.timeout = int(os.getenv('GOOGLE_TRENDS_TIMEOUT_MS', 90000)) / 1000
        self.concurrency = int(os.getenv('GOOGLE_TRENDS_CONCURRENCY', 1))

        # Proxy support (respects standard env vars or custom GOOGLE_TRENDS_PROXY)
        self.proxy = (
            os.getenv('GOOGLE_TRENDS_PROXY') or
            os.getenv('HTTPS_PROXY') or
            os.getenv('HTTP_PROXY') or
            ''
        )

        # Cookie support — paste browser cookies here to bypass 429 IP bans
        # Get from DevTools → Network → trends.google.com → request headers → Cookie
        raw_cookies = os.getenv('GOOGLE_TRENDS_COOKIES', '')
        self.cookies = self._parse_cookies(raw_cookies) if raw_cookies else {}

        # Lock for concurrency control (uses the module-level global lock)
        self.request_lock = _GLOBAL_REQUEST_LOCK
        self.pytrends = None  # Will be initialized per request
        
        if IS_TEST_MODE:
            logger.info('🧪 Google Trends Connector - MOCK MODE (No API calls will be made)')
        else:
            logger.info(
                f'🌐 Google Trends Connector - REAL API MODE (Enhanced Anti-block)\n'
                f'   Max Retries: {self.max_retries}\n'
                f'   Retry Delay: {self.retry_delay}s\n'
                f'   Request Delay: {self.request_delay}s\n'
                f'   Timeout: {self.timeout}s\n'
                f'   Concurrency: {self.concurrency}\n'
                f'   User Agents: {len(USER_AGENTS)} rotating agents'
            )
    
    def _parse_cookies(self, cookie_string: str) -> Dict[str, str]:
        """Parse a raw Cookie header string into a name→value dict.

        Non-latin-1 characters (e.g. the Unicode ellipsis '…' that DevTools
        inserts when truncating a long value) are stripped so that the
        requests library never hits a UnicodeEncodeError when building headers.
        """
        # Remove any character that HTTP headers cannot carry (must be latin-1)
        safe_string = cookie_string.encode('latin-1', errors='ignore').decode('latin-1')
        cookies: Dict[str, str] = {}
        for part in safe_string.split(';'):
            part = part.strip()
            if '=' in part:
                name, _, value = part.partition('=')
                cookies[name.strip()] = value.strip()
        if safe_string != cookie_string:
            logger.warning(
                '🍪 Cookie string contained non-latin-1 characters (likely DevTools truncation). '
                'Re-copy the full cookie value from DevTools for best results.'
            )
        return cookies

    def _get_fresh_client(self):
        """Create a fresh pytrends client with random user agent to avoid blocks.

        pytrends already calls GetGoogleCookie() in __init__, which visits
        trends.google.com/explore to obtain the NID cookie automatically.
        We just need to set realistic headers and inject any user-provided cookies.
        """
        user_agent = random.choice(USER_AGENTS)
        logger.debug(f'🔧 Creating fresh pytrends client')
        logger.debug(f'📱 User-Agent: {user_agent[:80]}...')

        # Mimic a real browser as closely as possible
        headers = {
            'User-Agent': user_agent,
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.9',
            'Accept-Encoding': 'gzip, deflate, br',
            'Connection': 'keep-alive',
            'DNT': '1',
            'Upgrade-Insecure-Requests': '1',
            'Sec-Fetch-Dest': 'document',
            'Sec-Fetch-Mode': 'navigate',
            'Sec-Fetch-Site': 'none',
            'Sec-Fetch-User': '?1',
        }

        requests_args: Dict = {'headers': headers}

        if self.proxy:
            requests_args['proxies'] = {'http': self.proxy, 'https': self.proxy}
            logger.debug(f'🌐 Routing through proxy: {self.proxy[:50]}')

        # TrendReq.__init__ calls GetGoogleCookie() which visits
        # trends.google.com/explore — so client.cookies already has NID on return.
        client = TrendReq(
            hl='en-US',
            tz=360,
            timeout=(self.timeout, self.timeout),
            retries=0,
            backoff_factor=0,
            requests_args=requests_args
        )

        nid_from_pytrends = 'NID' in (client.cookies or {})
        logger.debug(f'🍪 pytrends GetGoogleCookie NID present: {nid_from_pytrends} | cookies={list((client.cookies or {}).keys())}')

        # Inject user-provided cookies on top (override if same key)
        if self.cookies:
            client.cookies = {**(client.cookies or {}), **self.cookies}
            logger.debug(f'🍪 Merged user-provided cookies: {list(self.cookies.keys())}')

        return client
    
    def fetch_complete(
        self,
        keyword: str,
        country: str,
        window_days: int,
        baseline_days: int
    ) -> Dict:
        """
        Fetch complete trend data: time series + country comparison.
        
        Args:
            keyword: Search keyword
            country: ISO 2-letter country code (MX, CR, ES)
            window_days: Recent window for analysis
            baseline_days: Historical baseline period
            
        Returns:
            Dictionary with timeSeries and byCountry data
        """
        # Test mode: return mocks immediately
        if IS_TEST_MODE:
            time.sleep(0.005)  # Simulate minimal delay
            return {
                'timeSeries': generate_mock_time_series(keyword, baseline_days),
                'byCountry': generate_mock_by_country(keyword, country),
                'source': 'mock',
                'fetchedAt': datetime.utcnow().isoformat()
            }
        
        try:
            logger.info(
                f'🔍 Fetching complete trend data from Google Trends API',
                keyword=keyword,
                country=country,
                window_days=window_days,
                baseline_days=baseline_days
            )
            
            # Calculate date range
            end_date = datetime.now()
            start_date = end_date - timedelta(days=baseline_days)
            
            # Longer initial delay (5-10s) to appear more human and reduce IP bans
            initial_delay = 5 + random.random() * 5
            logger.info(f'⏳ Initial delay: {round(initial_delay, 1)}s (anti-bot measure)')
            time.sleep(initial_delay)

            # Create ONE shared client — GetGoogleCookie() is called only once per fetch_complete()
            client = self._get_fresh_client()
            logger.debug(f'🔑 Shared client NID present: {"NID" in (client.cookies or {})}')

            # Fetch time series with retry logic (reuses the same client)
            time_series = self._fetch_with_retry(
                lambda: self._fetch_time_series(keyword, country, start_date, end_date, client)
            )

            # Longer delay between the two calls (30-45s) — fresh quota window
            delay_between_requests = 30 + random.random() * 15
            logger.info(
                f'⏳ Waiting {round(delay_between_requests, 1)}s before fetching country data (rate-limit window)...'
            )
            time.sleep(delay_between_requests)

            # Fetch country comparison — graceful: if 429, return empty and keep time series
            by_country = []
            try:
                by_country = self._fetch_with_retry(
                    lambda: self._fetch_by_country(keyword, client)
                )
            except Exception as country_err:
                err_msg = str(country_err)
                is_rate_limit = '429' in err_msg or 'TooManyRequests' in type(country_err).__name__
                if is_rate_limit:
                    logger.warning(
                        f'⚠️  Country comparison skipped (Google 429 on by_region endpoint) '
                        f'— returning time series only. Try again in a few minutes.',
                        keyword=keyword
                    )
                else:
                    logger.warning(
                        f'⚠️  Country comparison failed ({type(country_err).__name__}: {err_msg[:200]}) '
                        f'— returning time series only.',
                        keyword=keyword
                    )

            logger.info(
                f'✅ Successfully fetched Google Trends data',
                keyword=keyword,
                country=country,
                data_points=len(time_series),
                by_country_available=len(by_country) > 0
            )

            return {
                'timeSeries': time_series,
                'byCountry': by_country,
                'source': 'google_trends',
                'fetchedAt': datetime.utcnow().isoformat()
            }
            
        except Exception as error:
            logger.error(
                f'❌ Failed to fetch Google Trends data',
                error=str(error),
                keyword=keyword,
                country=country
            )
            raise
    
    def _fetch_time_series(
        self,
        keyword: str,
        country: str,
        start_date: datetime,
        end_date: datetime,
        client: Optional[TrendReq] = None
    ) -> List[Dict]:
        """Fetch time series data for a keyword in a specific country."""
        logger.info(
            f'Fetching time series from Google Trends',
            keyword=keyword,
            country=country,
            start_date=start_date.strftime('%Y-%m-%d'),
            end_date=end_date.strftime('%Y-%m-%d')
        )

        pytrends = client or self._get_fresh_client()
        
        # Build payload for pytrends
        timeframe = f'{start_date.strftime("%Y-%m-%d")} {end_date.strftime("%Y-%m-%d")}'
        
        logger.debug(f'📦 Building payload: keyword={keyword}, timeframe={timeframe}, geo={country}')
        
        pytrends.build_payload(
            kw_list=[keyword],
            cat=0,
            timeframe=timeframe,
            geo=country,
            gprop=''
        )
        
        logger.debug(f'🌐 Calling pytrends.interest_over_time()...')
        
        # Get interest over time
        try:
            df = pytrends.interest_over_time()
            logger.debug(f'✅ Got response from Google Trends, checking data...')
        except Exception as e:
            logger.error(f'❌ Exception from pytrends: {type(e).__name__}: {str(e)[:500]}')
            raise
        
        if df.empty:
            logger.warning(f'No data returned from Google Trends for {keyword}')
            return []
        
        # Convert to list of dictionaries
        result = []
        for date_index, row in df.iterrows():
            result.append({
                'date': date_index.strftime('%Y-%m-%d'),
                'value': int(row[keyword]) if keyword in row else 0
            })
        
        logger.info(f'Successfully fetched time series ({len(result)} data points)')
        return result
    
    def _fetch_by_country(self, keyword: str, client: Optional[TrendReq] = None) -> List[Dict]:
        """Fetch country comparison data for a keyword (global, last 12 months)."""
        logger.info(f'Fetching country comparison from Google Trends', keyword=keyword)

        pytrends = client or self._get_fresh_client()

        # Log active cookie names (not values) so we can verify session state
        cookie_names = list(pytrends.cookies.keys()) if hasattr(pytrends, 'cookies') else []
        logger.debug(f'🍪 Active cookie names in session: {cookie_names}')

        timeframe = 'today 12-m'
        logger.debug(f'📦 Building payload: keyword={keyword}, timeframe={timeframe}, geo=(worldwide)')

        # Build payload for last 12 months, worldwide
        pytrends.build_payload(
            kw_list=[keyword],
            cat=0,
            timeframe=timeframe,
            geo='',
            gprop=''
        )

        logger.debug(f'🌐 Calling pytrends.interest_by_region()...')
        try:
            df = pytrends.interest_by_region(resolution='COUNTRY', inc_low_vol=True)
            logger.debug(f'✅ Got interest_by_region response — shape: {df.shape}, columns: {list(df.columns)}')
        except Exception as e:
            # Try to capture the HTTP response code if available
            resp_info = ''
            if hasattr(e, 'response') and e.response is not None:
                resp_info = f' | HTTP {e.response.status_code} | headers: {dict(e.response.headers)}'
            logger.error(f'❌ Exception from interest_by_region: {type(e).__name__}: {str(e)[:500]}{resp_info}')
            raise

        if df.empty:
            logger.warning(f'No country data returned from Google Trends for {keyword}')
            return []

        # Filter for supported countries and convert to list
        supported_countries = ['MX', 'CR', 'ES']
        result = []

        for country_code, row in df.iterrows():
            if country_code in supported_countries:
                result.append({
                    'country': country_code,
                    'value': int(row[keyword]) if keyword in row else 0
                })

        # Sort by value descending
        result.sort(key=lambda x: x['value'], reverse=True)

        logger.info(f'Successfully fetched country comparison ({len(result)} countries)')
        return result
    
    def _fetch_with_retry(self, fetch_fn):
        """
        Retry wrapper with exponential backoff.
        
        Args:
            fetch_fn: Function to execute (no arguments)
            
        Returns:
            Result from fetch_fn
            
        Raises:
            Exception after all retries exhausted
        """
        last_error = None

        for attempt in range(1, self.max_retries + 1):
            try:
                logger.debug(f'🔄 Attempt {attempt}/{self.max_retries} (waiting for global lock...)')
                with self.request_lock:
                    logger.debug(f'🔓 Lock acquired — running request')
                    return fetch_fn()
            except Exception as error:
                last_error = error

                # Detect if Google blocked with HTML response
                error_message = str(error)
                error_type = type(error).__name__

                # Log the FULL error for debugging
                logger.debug(f'🐛 Full error type: {error_type}')
                logger.debug(f'🐛 Full error message (first 1000 chars): {error_message[:1000]}')

                # Log HTTP response details and auto-refresh NID cookie from Set-Cookie header
                if hasattr(error, 'response') and error.response is not None:
                    r = error.response
                    retry_after = r.headers.get('Retry-After', 'not set')
                    logger.debug(
                        f'🔎 HTTP response details | status={r.status_code} '
                        f'| Retry-After={retry_after} '
                        f'| Content-Type={r.headers.get("Content-Type", "?")} '
                        f'| body_preview={r.text[:300]!r}'
                    )
                    # Google often sends a fresh NID in the 429 Set-Cookie — use it next attempt
                    set_cookie = r.headers.get('Set-Cookie', '')
                    nid_match = re.search(r'NID=([^;]+)', set_cookie)
                    if nid_match:
                        new_nid = nid_match.group(1)
                        self.cookies['NID'] = new_nid
                        logger.info(f'🍪 Auto-refreshed NID cookie from 429 response (will use on next attempt)')

                is_rate_limited = 'TooManyRequests' in error_type or '429' in error_message
                is_blocked = (
                    is_rate_limited or
                    'Unexpected token' in error_message or
                    'is not valid JSON' in error_message or
                    '<html' in error_message.lower() or
                    '<!doctype' in error_message.lower() or
                    'Invalid JSON' in error_message or
                    'The request failed' in error_message or
                    'JSONDecodeError' in error_type
                )

                if attempt < self.max_retries:
                    # Exponential backoff con factor aleatorio: base * 1.5^attempt + random(0-5s)
                    # Para 429 usa un delay extra largo (mínimo 30s)
                    base_delay = max(self.retry_delay, 30) if is_rate_limited else self.retry_delay
                    delay_time = round(base_delay * math.pow(1.5, attempt - 1) + random.random() * 5, 1)

                    logger.warning(
                        f'{"🚫 Google rate-limited (429)" if is_rate_limited else ("🚫 Google BLOCKED request" if is_blocked else "⚠️  Request failed")} '
                        f'- waiting {delay_time}s before retry...',
                        attempt=attempt,
                        max_retries=self.max_retries,
                        error_type=error_type,
                        error=error_message[:150],
                        is_blocked=is_blocked,
                        next_delay_seconds=delay_time
                    )

                    time.sleep(delay_time)
                else:
                    logger.error(
                        f'❌ All retry attempts exhausted',
                        attempt=attempt,
                        max_retries=self.max_retries,
                        error_type=error_type,
                        error=error_message[:500]
                    )

        raise Exception(f'Failed after {self.max_retries} attempts: {str(last_error)}')


# Singleton instance
google_trends_connector = GoogleTrendsConnector()


# Export for testing
__all__ = [
    'GoogleTrendsConnector',
    'google_trends_connector',
    'generate_mock_time_series',
    'generate_mock_by_country'
]
